spec/requests/spree/admin/payments_spec.rb

```diff
@@ -10,6 +10,68 @@
     sign_in create(:admin_user)
   end
 
+  describe "POST /admin/orders/:order_number/payments.json" do
+    let(:params) do
+      {
+        payment: {
+          payment_method_id: payment_method.id, amount: order.total
+        }
+      }
+    end
+    let(:payment_method) do
+      create(:payment_method, distributors: [order.distributor])
+    end
+
+    it "creates a payment" do
+      expect {
+        post("/admin/orders/#{order.number}/payments.json", params:)
+      }.to change { order.payments.count }.by(1)
+    end
+
+    it "redirect to payments page" do
+      post("/admin/orders/#{order.number}/payments.json", params:)
+
+      expect(response).to redirect_to(spree.admin_order_payments_path(order))
+      expect(flash[:success]).to eq "Payment has been successfully created!"
+    end
+
+    context "when failing to create payment" do
+      it "redirects to payments page" do
+        payment_mock = instance_double(Spree::Payment)
+        allow(order.payments).to receive(:build).and_return(payment_mock)
+        allow(payment_mock).to receive(:save).and_return(false)
```

Comment:
I think it's better practice to make an invalid payment. If there's no way to make  payment validation fail, then maybe we don't need to test this 🙂

Reply:
I don't think there is an easy way to make an invalid payment. How would you test this then ?
https://github.com/openfoodfoundation/openfoodnetwork/blob/5d9bb9a8d5ac2b491bc95d6fae1bab24c4f8c7de/app/controllers/spree/admin/payments_controller.rb#L31-L34
I don't know enough about how this is used to assume `@paymen.save` doesn't fail, so I'd rather have a test for it.

Reply:
Looking at the model, it validates the payment source. A Spree::CreditCard can be invalid if the month, year or number is empty.

Requires a bit more setup. I guess you can argue more practical and more efficient to mock it, as you have ;)